### DISCUSSION TITLE
Suggestion to rename PRBProxyAnnex

### DISCUSSION THREAD

**Discussion by razgraf** [PaulRBerg/prb-proxy#106] [Ideas]
A recent update included a refactor of the `Helpers` contract into `PRBProxyAnnex`. The new name doesn't really imply anything about the purpose of the contract, other than it being an additional functionality package.

### Topic 1: Rename it

Would suggest renaming it to something else. Possibly:
1. PRBProxyExtension (implying it extends functionality through plugins)
2. PRBProxyInject (implying it does handle dependency injection - which is what the plugins do [manage extra functionality])

### Topic 2: Make the "target" classification more explicit

This contract is basically a target contract for the proxy. Moreover, it lives on a different layer than Proxy and Registry which could be bundled under the "factory" purpose - creator and item.

How about either:
1. Adding a "target" particle in the name of this contract
2. Creating a "target" folder and placing it inside

The 2nd possibility has the added benefit of being scalable long-term. The next time you add a system target or simply a popular target many users desire, you can just drop it in there.

**Comment by PaulRBerg** [comment]
Thanks for your feedback.

> The new name doesn't really imply anything about the purpose of the contract

It definitely implies something; just like any word does. I based my choice on this definition I saw on [dictionary.com](https://www.dictionary.com/browse/annex):

"a subsidiary building or an addition to a building"

>  other than it being an additional functionality package.

This is exactly what it is.

> PRBProxyExtension

Good name, but since the audits have already started, I would be reluctant to make this name change if that would mean paying extra.

> (implying it extends functionality through plugins)

But the `Annex` doesn't do this. It is the user and the plugins together who extend the functionality of the proxy; not the `Annex`. The `Annex` is just an ancillary component that gets delegate-called.

> implying it does handle dependency injection

I want to steer away from this terminology because it creates a lot of unnecessary confusion, and it's simply not the right word for the job.

Target contracts are not dependencies; the proxy does not "depend" upon targets for anything. Targets are collections of optional stateless scripts that can be swapped out at any moment. Quite the opposite of dependency.

> which is what the plugins do [manage extra functionality])

Extra functionality != dependencies. The proxy is not dependent on any plugin.

Also, the `Annex` has more logic than the proxy install and uninstall methods.

> Adding a "target" particle in the name of this contract

Yeah so one of the first ideas I had for this contract's name was `EnshrinedTarget`. But I wanted to keep the name short (one word), for brevity and simplicity overall.

> Creating a "target" folder and placing it inside

1. Overkill for one target contract. Generally in software, we don't create directories for single files.
2. Etherscan users would nonetheless not see the `target/` path.

> The 2nd possibility has the added benefit of being scalable long-term

I don't plan on further developing the proxy after we launch Sablier V2; we will be stuck on the current version (after the audit finishes) for a long time - migrating proxies involves a lot of work.

> popular target many users desire

These should live in a separate repo to avoid turning the core PRBProxy repo into a bazaar of features.

**Reply by razgraf** [reply]
> Target contracts are not dependencies; the proxy does not "depend" upon targets for anything.

I think you are disputing the meaning of the sequence of words "dependency injection" instead of focusing on the concept of dependency injection. Agree that an entity of type **dependency** is not the same as one of type **plugin**, but the injection mechanism of this target contract works exactly as described by the aforementioned concept.

**Reply by PaulRBerg** [reply]
I don't follow your point. Could you please restate it? And also could you follow up on the other points I have made? I'm not sure if you agree or disagree with them.

First, I am confused by your first sentence because "meaning of sequence of words" = concept.

Second, if you agree that dependency != plugin, I really don't see why we should mudden the terminology of PRBProxy with smth always referred to in the context of dependencies. Like why do this, when there are alternatives that don't come with this baggage?

Finally, in the general case: even if we did in fact talk about contract dependencies, I would still be reluctant to use the word "injection", because contracts cannot be injected in the traditional software engineering sense. They can be "delegate called"; that's the precise term. It's a new paradigm.

**Reply by razgraf** [reply]
> It definitely implies something; just like any word does

The quote you selected lacks context. Reading the entire phrase, you can see I've stated "imply anything **about**" - which reduces the contextual implications from the general definition of the word to written structures that define the element they name more precisely. It's like choosing between "Contract.sol" and "PRBProxy" when naming the "PRBProxy.sol" contract.

> "a subsidiary building or an addition to a building"

If we're going after specific definitions, what do buildings have to do with contracts. Maybe you meant to select the definition right next to it: "something added to a document; appendix; supplement".

> I want to steer away from this terminology because it creates a lot of unnecessary confusion

Sure thing, it's your contract. What I have been trying to say is that there should be little to no confusion because what you do with the plugins can literally be called "injecting functionality" - the closest well known and similar concept to it, that we as programmers should know is "dependency injection".

> Also, the Annex has more logic than the proxy install and uninstall methods.

That is a very good point. Meaning the word "inject" may not do the `permissions` and `gasReserve` functionalities justice. I can agree with that.

> Generally in software, we don't create directories for single files

I've simply suggested that in terms of scalability for the future. If the plan is not to expand this product in the future, I can understand that.

The intent was not to turn this repository into a bazaar although it could be a great idea to provide another `prb-proxy-targets` repository where people can code and share their own implementations for the public to see.

> Etherscan users would nonetheless not see the target/ path.

I'm sure programmers/integrators would gladly use github rather than **only** etherscan to understand contracts they may depend upon.

>  I'm not sure if you agree or disagree with them.

It's a bit tiring to re-explain a point when you're picking it apart rather than taking it as a whole. Not that doing it is inherently wrong, just a tad bit tiring at least for me, now.

Not answering is not a form of agreeing or disagreeing, it's just that I don't consider it worth it to spend more time on it. Once I've provided my opinion I'm more than happy to accept that it's wrong or that some items are worthy of critic - while at the same time I don't exactly see the value in committing more time to it if you've already made up your mind.

> I don't follow your point. Could you please restate it?

I've stated it pretty clearly. Instead of focusing on the value a dependency injection (or functionality injection) concept brings to the table, you got stuck on the word "dependency".

> I am confused by your first sentence "meaning of sequence of words" = concept.

Sequence was a poor choice of words as it implies an order. Replace sequence with "list of words, interpreted individually". In short, "dependency" and "injection" may relate to one thing when taken individually and another when taken together (the concept).

> if you agree that dependency != plugin

They're different in the strong sense of the words, but I don't see acknowledging plugins as soft/optional-dependencies as something impossible.

>  It's a new paradigm.

Nothing bad in matching/comparing new paradigms with parts of old and reliable ones.

---

TL;DR; I see the point you're trying to make with "dependencies" not being the right words, as well as "injection" being too close to it from a programmer's pov.

Overall, would still be nice to see this contract bearing a different name. That is what the OP is suggesting.

**Reply by PaulRBerg** [reply]
> The quote you selected lacks context. Reading the entire phrase, you can see I've stated "imply anything about"

"imply" is a transitive verb, so is is equivalent to "imply anything about". Imply without an object is meaningless and not proper English.

> If we're going after specific definitions, what do buildings have to do with contracts

Metaphors and analogies are commonplace in software engineering.

> Maybe you meant to select the definition right next to it: "something added to a document; appendix; supplement".

Even better!

> Sure thing, it's your contract

The reason I want to steer way from the "dependency" terminology is because it's not helpful in an objective way.

> with the plugins can literally be called "injecting functionality

1. Moving the goal posts from "dependency injection" to "functionality injection".
2. At any rate, `Annex` is not just about plugins.
3. At any rate again, the way `DELEGATECALL` works is quite the other way around; the proxy state and address are "injected" in the logic implemented in `Annex`.

> That is a very good point. Meaning the word "inject" may not do the `permissions` and `gasReserve` functionalities justice

Thank you.

> it could be a great idea to provide another `prb-proxy-targets` repository

Yep, this would be nice.

> It's a bit tiring to re-explain a point when you're picking it apart rather than taking it as a whole

Software engineering is tiring. It's not about us as persons - this stuff is hard. But anyway, point well taken, I understand we've limited bandwidth.

> while at the same time I don't exactly see the value in committing more time to it if you've already made up your mind

The value is in finding refutations.

I'm always eager to change my mind given the right (objective) explanation.

> Instead of focusing on the value a dependency injection (or functionality injection) concept brings to the table, you got stuck on the word "dependency".

- dependency injection != functionality injection. Qualitatively different ideas.
- Moving the goal post from multi-word concepts to single words (previously you said "meaning of sequence of words").

> I've stated it pretty clearly.
> ...
> Sequence was a poor choice of words as it implies an order

These two statements are contradictory.

> I don't see acknowledging plugins as soft/optional-dependencies as something impossible.

Not impossible; just not helpful when there are alternative wordings like `Annex` and `Extension`.

As I've said, I will look into using `Extension`; which is a great choice.

> I see the point you're trying to make with "dependencies" not being the right words, as well as "injection" being too close to it from a programmer's pov.

Great, thanks!

**Comment by PaulRBerg** [comment]
On a recent call, @andreivladbrg made a good point against renaming to `Extension` - the proxy contract itself is an extension of the owner EOA. And his statement is all the more valid in the context of [removing the `transferOwnership` function](https://github.com/PaulRBerg/prb-proxy/issues/108), since the proxy would become a life-long extension of the EOA (an "extended arm" if you will).

I continue to agree that `Annex` is not super duper clear, but I also think it's the least bad option (per the refutation above that `Inject` is misleading; the flow of "injection" is the other way around).

_Side note_: the inherent computational universality of the proxy (i.e., [`execute`](https://github.com/PaulRBerg/prb-proxy/blob/6fbb6df62d25e07fc702a4485595aa34d8297237/src/PRBProxy.sol#L87)) makes it difficult to draw neat categories for it.

**Comment by PaulRBerg** [comment]
As faith would have it, I ended up removing the `Annex` contract because this was a stepping stone for fixing [one of Cantina's findings]( https://github.com/cantinasec/review-sablier2/issues/13):

https://github.com/PaulRBerg/prb-proxy/pull/120

Gonna close this discussion as "outdated".
